File: agentic_os/mission/events.py
"""Normalized runtime events — the public event vocabulary + local, replayable event store.

The load-bearing idea (from the cross-agent-observability model): **every supported agent surface, live
or persisted, normalizes into the same typed, replayable event history.** Python is the executable
specification; this mirrors the v9 amendment's canonical event schemas and the v10 ``RuntimeEvent`` family.

This module is the public contract:

* one versioned :class:`RuntimeEvent` envelope (content-addressed id → deterministic, replayable) +
  thin specialization constructors;
* the versioned per-record :data:`SCHEMAS` and :meth:`RuntimeEvent.validate`;
* an append-only, physically-ordered :class:`EventLedger` with causal edges + NDJSON serialization —
  a minimal *local* event store, enough to record, replay, audit and conform.

What is NOT here (enterprise overlay): fleet-wide **enforcement** (the clean-decision Decision Gate),
**policy correlation** (the Trajectory rule engine), enterprise event ingestion, forensic storage,
cross-tenant analytics, and operational console aggregation. Those consume this same vocabulary; a
public-only deployment records and replays events without them.

Contract version: :data:`SCHEMA_VERSION` (``runtime-event/v1``). Additive-only within a major (new event
types, new optional envelope fields, new specialized schemas); a breaking change to the envelope or the
NDJSON serialization bumps the major. The ``_demo`` self-test is the executable golden fixture.
"""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from enum import Enum
from hashlib import sha256


# ════════════════════════════ Runtime Event Contract ════════════════════════════
class EventType(str, Enum):
    DEREFERENCE = "dereference"; CAPABILITY_INVOCATION = "capability_invocation"
    TOOL_PROPOSAL = "tool_proposal"; TOOL_RESULT = "tool_result"; VERIFICATION = "verification"
    MERGE = "merge"; PROMOTION = "promotion"; POLICY_DECISION = "policy_decision"
    HUMAN_REVIEW = "human_review"


class ResultStatus(str, Enum):
    PROPOSED = "proposed"; ATTEMPTED = "attempted"; COMPLETED = "completed"
    FAILED = "failed"; OBSERVED = "observed"; DENIED = "denied"


class Severity(str, Enum):
    INFORMATIONAL = "informational"; LOW = "low"; MEDIUM = "medium"; HIGH = "high"; CRITICAL = "critical"

    @property
    def rank(self) -> int:
        return ["informational", "low", "medium", "high", "critical"].index(self.value)


#: Semantic version of the public runtime-event contract (the envelope + NDJSON serialization).
SCHEMA_VERSION = "runtime-event/v1"
CONTRACT_VERSION = SCHEMA_VERSION

# the versioned schemas the v9 amendment enumerates (required fields per record kind)
SCHEMAS: dict[str, list[str]] = {
    "runtime-event/v1": ["event_id", "event_type", "schema_version", "timestamp", "actor", "result_status"],
    "dereference-event/v1": ["event_id", "artifact_handles", "visibility"],
    "capability-event/v1": ["event_id", "capability_id"],
    "verification-event/v1": ["event_id", "result_status"],
    "policy-decision/v1": ["event_id", "policy_context"],
    "session-summary/v1": ["session_id", "event_count"],
}


@dataclass
class RuntimeEvent:
    """The one normalized envelope every subsystem emits — Context/Mission/Discovery Runtimes,
    Sidekick, and external-agent adapters share this single forensic timeline."""
    event_type: EventType
    actor: str
    timestamp: int                                  # caller-supplied logical time (deterministic/replayable)
    schema_version: str = SCHEMA_VERSION
    mission_id: str = ""
    session_id: str = ""
    parent_event_id: str = ""
    capability_id: str = ""
    artifact_handles: list[str] = field(default_factory=list)
    source_runtime: str = ""
    source_agent: str = ""
    visibility: str = "private"
    policy_context: str = ""
    result_status: ResultStatus = ResultStatus.OBSERVED
    evidence_refs: list[str] = field(default_factory=list)
    payload: dict = field(default_factory=dict)
    event_id: str = ""

    def __post_init__(self) -> None:
        if not self.event_id:                       # content-addressed id → deterministic, replayable
            key = f"{self.event_type.value}|{self.actor}|{self.timestamp}|{self.mission_id}|{self.parent_event_id}|{sorted(self.artifact_handles)}|{json.dumps(self.payload, sort_keys=True)}"
            self.event_id = "ev-" + sha256(key.encode()).hexdigest()[:16]

    def to_ndjson(self) -> str:
        d = asdict(self)
        d["event_type"] = self.event_type.value
        d["result_status"] = self.result_status.value
        return json.dumps(d, sort_keys=True)

    @classmethod
    def from_ndjson(cls, line: str) -> "RuntimeEvent":
        d = json.loads(line)
        d["event_type"] = EventType(d["event_type"])
        d["result_status"] = ResultStatus(d["result_status"])
        return cls(**d)

    def validate(self) -> list[str]:
        """Missing required fields per the envelope + the type's specialized schema (empty = valid)."""
        spec_schema = {
            EventType.DEREFERENCE: "dereference-event/v1", EventType.CAPABILITY_INVOCATION: "capability-event/v1",
            EventType.VERIFICATION: "verification-event/v1", EventType.POLICY_DECISION: "policy-decision/v1",
        }.get(self.event_type)
        d = asdict(self)
        missing = [f for f in SCHEMAS["runtime-event/v1"] if not d.get(f)]
        if spec_schema:
            missing += [f for f in SCHEMAS[spec_schema] if not d.get(f)]
        return missing


# specialization constructors (thin — the envelope is the contract)
def dereference_event(actor, timestamp, artifact_handles, visibility="private", **kw) -> RuntimeEvent:
    return RuntimeEvent(EventType.DEREFERENCE, actor, timestamp, artifact_handles=artifact_handles, visibility=visibility, **kw)


def capability_event(actor, timestamp, capability_id, result_status=ResultStatus.ATTEMPTED, **kw) -> RuntimeEvent:
    return RuntimeEvent(EventType.CAPABILITY_INVOCATION, actor, timestamp, capability_id=capability_id, result_status=result_status, **kw)


def tool_result_event(actor, timestamp, capability_id, result_status=ResultStatus.COMPLETED, **kw) -> RuntimeEvent:
    return RuntimeEvent(EventType.TOOL_RESULT, actor, timestamp, capability_id=capability_id, result_status=result_status, **kw)


def policy_decision_event(actor, timestamp, policy_context, result_status, **kw) -> RuntimeEvent:
    return RuntimeEvent(EventType.POLICY_DECISION, actor, timestamp, policy_context=policy_context, result_status=result_status, **kw)


class EventLedger:
    """Append-only, physically-ordered event history with causal (parent) edges. Never mutates or
    erases — a model rewind, an abandoned branch, and a superseded proposal are all retained. This is the
    minimal *local* store; enterprise ingestion/forensics/analytics consume the same NDJSON."""

    def __init__(self) -> None:
        self._events: list[RuntimeEvent] = []
        self._by_id: dict[str, RuntimeEvent] = {}

    def append(self, event: RuntimeEvent) -> RuntimeEvent:
        self._events.append(event)                  # physical order preserved; never rewritten
        self._by_id[event.event_id] = event
        return event

    def events(self) -> list[RuntimeEvent]:
        return list(self._events)

    def get(self, event_id: str) -> RuntimeEvent | None:
        return self._by_id.get(event_id)

    def children(self, event_id: str) -> list[RuntimeEvent]:
        return [e for e in self._events if e.parent_event_id == event_id]

    def by_session(self, session_id: str) -> list[RuntimeEvent]:
        return [e for e in self._events if e.session_id == session_id]

    def to_ndjson(self) -> str:
        return "\n".join(e.to_ndjson() for e in self._events)

    @classmethod
    def from_ndjson(cls, text: str) -> "EventLedger":
        led = cls()
        for line in text.splitlines():
            if line.strip():
                led.append(RuntimeEvent.from_ndjson(line))
        return led

    def session_summary(self, session_id: str) -> dict:
        evs = self.by_session(session_id)
        return {"schema_version": "session-summary/v1", "session_id": session_id, "event_count": len(evs),
                "types": sorted({e.event_type.value for e in evs}),
                "failed": sum(1 for e in evs if e.result_status is ResultStatus.FAILED)}


# ════════════════════════════ self-test / golden fixture ════════════════════════════
def _demo() -> int:
    checks: list[tuple[str, bool]] = []

    # content-addressed id is deterministic (same inputs → same id → replayable)
    e1 = capability_event("agent", 1, "cap.read", mission_id="m1")
    e1b = capability_event("agent", 1, "cap.read", mission_id="m1")
    checks.append(("content-addressed event_id is deterministic", e1.event_id == e1b.event_id))

    # NDJSON round-trips byte-stably
    checks.append(("RuntimeEvent NDJSON round-trips", RuntimeEvent.from_ndjson(e1.to_ndjson()).to_ndjson() == e1.to_ndjson()))

    # validate flags a missing required field on a specialized schema
    bad = RuntimeEvent(EventType.DEREFERENCE, "agent", 2)   # no artifact_handles
    checks.append(("validate catches a missing specialized field", "artifact_handles" in bad.validate()))
    good = dereference_event("agent", 3, ["art-1"])
    checks.append(("a complete event validates clean", good.validate() == []))

    # ledger preserves physical order + causal edges, and replays from NDJSON
    led = EventLedger()
    p = led.append(capability_event("agent", 4, "cap.plan", session_id="s1"))
    led.append(tool_result_event("agent", 5, "cap.plan", session_id="s1", parent_event_id=p.event_id))
    replay = EventLedger.from_ndjson(led.to_ndjson())
    checks.append(("ledger replays from NDJSON identically", replay.to_ndjson() == led.to_ndjson()))
    checks.append(("causal children resolve", len(led.children(p.event_id)) == 1))
    checks.append(("session summary counts + types", led.session_summary("s1")["event_count"] == 2))

    for name, ok in checks:
        print(f"  check {name} = {ok}")
    passed = sum(1 for _, ok in checks if ok)
    print(f"RESULT {passed}/{len(checks)}  (event contract {SCHEMA_VERSION})")
    return 0 if passed == len(checks) else 1


if __name__ == "__main__":
    import sys
    sys.exit(_demo())

File: go/mission/events.go
package mission

// Normalized runtime events — the public event vocabulary + a local, replayable event store.
//
// Go port of agentic_os/mission/events.py (the executable specification). Every supported agent
// surface normalizes into the same typed, replayable event history: one content-addressed
// RuntimeEvent envelope (deterministic id → replayable), the versioned per-record SCHEMAS +
// Validate, and an append-only EventLedger with causal edges + NDJSON serialization.
//
// The content-addressed event_id and the NDJSON encoding are BIT-IDENTICAL to the Python (and
// Kotlin) implementations: the id key embeds Python's list repr for the sorted artifact handles and
// json.dumps(payload, sort_keys=True); NDJSON is json.dumps(dict, sort_keys=True) — both with
// Python's default ", " / ": " separators. pyJSON + pyListRepr below reproduce that exactly, so an
// event recorded on one runtime is verifiable on another. Contract version SchemaVersion below.

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// EventType is the public event vocabulary.
type EventType string

const (
	EvDereference          EventType = "dereference"
	EvCapabilityInvocation EventType = "capability_invocation"
	EvToolProposal         EventType = "tool_proposal"
	EvToolResult           EventType = "tool_result"
	EvVerification         EventType = "verification"
	EvMerge                EventType = "merge"
	EvPromotion            EventType = "promotion"
	EvPolicyDecision       EventType = "policy_decision"
	EvHumanReview          EventType = "human_review"
	// EvEvidenceChange — a world/evidence delta on the governance timeline (v0.2.x Slice 4), so
	// Governance can correlate an evidence-change series against an action series.
	EvEvidenceChange EventType = "evidence_change"
)

// ResultStatus is the outcome of the event's action.
type ResultStatus string

const (
	RsProposed  ResultStatus = "proposed"
	RsAttempted ResultStatus = "attempted"
	RsCompleted ResultStatus = "completed"
	RsFailed    ResultStatus = "failed"
	RsObserved  ResultStatus = "observed"
	RsDenied    ResultStatus = "denied"
)

// Severity ranks an event's importance (informational < low < medium < high < critical).
type Severity string

const (
	SevInformational Severity = "informational"
	SevLow           Severity = "low"
	SevMedium        Severity = "medium"
	SevHigh          Severity = "high"
	SevCritical      Severity = "critical"
)

var severityOrder = []Severity{SevInformational, SevLow, SevMedium, SevHigh, SevCritical}

// Rank returns the ordinal of a severity (-1 if unknown).
func (s Severity) Rank() int {
	for i, v := range severityOrder {
		if v == s {
			return i
		}
	}
	return -1
}

// SchemaVersion is the semantic version of the public runtime-event contract (envelope + NDJSON).
const SchemaVersion = "runtime-event/v10"

// ContractVersion is an alias kept for parity with the Python module surface.
const ContractVersion = SchemaVersion

// SCHEMAS enumerates the required fields per record kind (the v9 amendment's versioned schemas).
var SCHEMAS = map[string][]string{
	"runtime-event/v10":     {"event_id", "event_type", "schema_version", "timestamp", "actor", "result_status"},
	"dereference-event/v1":  {"event_id", "artifact_handles", "visibility"},
	"capability-event/v1":   {"event_id", "capability_id"},
	"verification-event/v1": {"event_id", "result_status"},
	"policy-decision/v1":    {"event_id", "policy_context"},
	"session-summary/v1":    {"session_id", "event_count"},
}

// RuntimeEvent is the one normalized envelope every subsystem emits — Context/Mission/Discovery
// Runtimes, Sidekick and external-agent adapters share this single forensic timeline.
type RuntimeEvent struct {
	EventType       EventType      `json:"event_type"`
	Actor           string         `json:"actor"`
	Timestamp       int            `json:"timestamp"` // caller-supplied logical time (deterministic/replayable)
	SchemaVersion   string         `json:"schema_version"`
	MissionID       string         `json:"mission_id"`
	SessionID       string         `json:"session_id"`
	ParentEventID   string         `json:"parent_event_id"`
	CapabilityID    string         `json:"capability_id"`
	ArtifactHandles []string       `json:"artifact_handles"`
	SourceRuntime   string         `json:"source_runtime"`
	SourceAgent     string         `json:"source_agent"`
	Visibility      string         `json:"visibility"`
	PolicyContext   string         `json:"policy_context"`
	ResultStatus    ResultStatus   `json:"result_status"`
	EvidenceRefs    []string       `json:"evidence_refs"`
	Payload         map[string]any `json:"payload"`
	EventID         string         `json:"event_id"`
}

// finalize applies the envelope defaults and stamps the content-addressed id (mirrors
// RuntimeEvent.__post_init__). Called by every constructor; keep it the single id authority.
func (e *RuntimeEvent) finalize() {
	if e.SchemaVersion == "" {
		e.SchemaVersion = SchemaVersion
	}
	if e.ResultStatus == "" {
		e.ResultStatus = RsObserved
	}
	if e.Visibility == "" {
		e.Visibility = "private"
	}
	if e.ArtifactHandles == nil {
		e.ArtifactHandles = []string{}
	}
	if e.EvidenceRefs == nil {
		e.EvidenceRefs = []string{}
	}
	if e.Payload == nil {
		e.Payload = map[string]any{}
	}
	if e.EventID == "" { // content-addressed id → deterministic, replayable
		handles := append([]string{}, e.ArtifactHandles...)
		sort.Strings(handles)
		key := strings.Join([]string{
			string(e.EventType), e.Actor, strconv.Itoa(e.Timestamp), e.MissionID,
			e.ParentEventID, pyListRepr(handles), pyJSON(e.Payload),
		}, "|")
		sum := sha256.Sum256([]byte(key))
		e.EventID = "ev-" + fmt.Sprintf("%x", sum)[:16]
	}
}

// NewRuntimeEvent builds an event with defaults applied and its content-addressed id stamped.
func NewRuntimeEvent(e RuntimeEvent) RuntimeEvent {
	e.finalize()
	return e
}

// ndjsonMap projects the event to the ordered field set json.dumps serializes (enum values as
// their strings), so ToNDJSON reproduces the Python bytes exactly.
func (e RuntimeEvent) ndjsonMap() map[string]any {
	return map[string]any{
		"event_type":       string(e.EventType),
		"actor":            e.Actor,
		"timestamp":        e.Timestamp,
		"schema_version":   e.SchemaVersion,
		"mission_id":       e.MissionID,
		"session_id":       e.SessionID,
		"parent_event_id":  e.ParentEventID,
		"capability_id":    e.CapabilityID,
		"artifact_handles": stringsToAny(e.ArtifactHandles),
		"source_runtime":   e.SourceRuntime,
		"source_agent":     e.SourceAgent,
		"visibility":       e.Visibility,
		"policy_context":   e.PolicyContext,
		"result_status":    string(e.ResultStatus),
		"evidence_refs":    stringsToAny(e.EvidenceRefs),
		"payload":          e.Payload,
		"event_id":         e.EventID,
	}
}

// ToNDJSON serializes the event as json.dumps(asdict, sort_keys=True) — Python's default separators.
func (e RuntimeEvent) ToNDJSON() string { return pyJSON(e.ndjsonMap()) }

// RuntimeEventFromNDJSON reconstructs an event from one NDJSON line. The event_id is carried
// verbatim (never recomputed), so a round-trip is byte-stable.
func RuntimeEventFromNDJSON(line string) (RuntimeEvent, error) {
	var raw map[string]any
	dec := json.NewDecoder(strings.NewReader(line))
	dec.UseNumber()
	if err := dec.Decode(&raw); err != nil {
		return RuntimeEvent{}, err
	}
	ev := RuntimeEvent{
		EventType:       EventType(asString(raw["event_type"])),
		Actor:           asString(raw["actor"]),
		Timestamp:       asInt(raw["timestamp"]),
		SchemaVersion:   asString(raw["schema_version"]),
		MissionID:       asString(raw["mission_id"]),
		SessionID:       asString(raw["session_id"]),
		ParentEventID:   asString(raw["parent_event_id"]),
		CapabilityID:    asString(raw["capability_id"]),
		ArtifactHandles: anyToStrings(raw["artifact_handles"]),
		SourceRuntime:   asString(raw["source_runtime"]),
		SourceAgent:     asString(raw["source_agent"]),
		Visibility:      asString(raw["visibility"]),
		PolicyContext:   asString(raw["policy_context"]),
		ResultStatus:    ResultStatus(asString(raw["result_status"])),
		EvidenceRefs:    anyToStrings(raw["evidence_refs"]),
		Payload:         asMap(raw["payload"]),
		EventID:         asString(raw["event_id"]), // preserved, not recomputed
	}
	if ev.Payload == nil {
		ev.Payload = map[string]any{}
	}
	return ev, nil
}

// Validate returns the missing required fields per the envelope + the type's specialized schema
// (empty slice = valid). Mirrors RuntimeEvent.validate.
func (e RuntimeEvent) Validate() []string {
	present := e.ndjsonMap()
	// event_id is required and always populated after finalize; a zero envelope still reports gaps.
	miss := func(fields []string) []string {
		var out []string
		for _, f := range fields {
			if isEmptyField(present[f]) {
				out = append(out, f)
			}
		}
		return out
	}
	missing := miss(SCHEMAS["runtime-event/v10"])
	specSchema := map[EventType]string{
		EvDereference:          "dereference-event/v1",
		EvCapabilityInvocation: "capability-event/v1",
		EvVerification:         "verification-event/v1",
		EvPolicyDecision:       "policy-decision/v1",
	}[e.EventType]
	if specSchema != "" {
		missing = append(missing, miss(SCHEMAS[specSchema])...)
	}
	return missing
}

// ─── specialization constructors (thin — the envelope is the contract) ───────

// DereferenceEvent records an artifact dereference (which handles, at what visibility).
func DereferenceEvent(actor string, timestamp int, artifactHandles []string, visibility string, base RuntimeEvent) RuntimeEvent {
	base.EventType, base.Actor, base.Timestamp = EvDereference, actor, timestamp
	base.ArtifactHandles = artifactHandles
	if visibility != "" {
		base.Visibility = visibility
	}
	return NewRuntimeEvent(base)
}

// CapabilityEvent records a capability invocation (defaults to ATTEMPTED).
func CapabilityEvent(actor string, timestamp int, capabilityID string, base RuntimeEvent) RuntimeEvent {
	base.EventType, base.Actor, base.Timestamp = EvCapabilityInvocation, actor, timestamp
	base.CapabilityID = capabilityID
	if base.ResultStatus == "" {
		base.ResultStatus = RsAttempted
	}
	return NewRuntimeEvent(base)
}

// ToolResultEvent records a tool result (defaults to COMPLETED).
func ToolResultEvent(actor string, timestamp int, capabilityID string, base RuntimeEvent) RuntimeEvent {
	base.EventType, base.Actor, base.Timestamp = EvToolResult, actor, timestamp
	base.CapabilityID = capabilityID
	if base.ResultStatus == "" {
		base.ResultStatus = RsCompleted
	}
	return NewRuntimeEvent(base)
}

// PolicyDecisionEvent records a governance decision on the timeline.
func PolicyDecisionEvent(actor string, timestamp int, policyContext string, resultStatus ResultStatus, base RuntimeEvent) RuntimeEvent {
	base.EventType, base.Actor, base.Timestamp = EvPolicyDecision, actor, timestamp
	base.PolicyContext, base.ResultStatus = policyContext, resultStatus
	return NewRuntimeEvent(base)
}

// EvidenceChangeEvent records a world/evidence delta (v0.2.x Slice 4): ref is the logical artifact
// id and changeType is created|updated|deleted; both land in the payload and evidence_refs so a
// trajectory rule can correlate the evidence series with a following action series.
func EvidenceChangeEvent(actor string, timestamp int, ref, changeType string, base RuntimeEvent) RuntimeEvent {
	base.EventType, base.Actor, base.Timestamp = EvEvidenceChange, actor, timestamp
	payload := map[string]any{}
	for k, v := range base.Payload {
		payload[k] = v
	}
	payload["ref"], payload["change_type"] = ref, changeType
	base.Payload = payload
	base.EvidenceRefs = []string{ref}
	base.ResultStatus = RsObserved
	return NewRuntimeEvent(base)
}

// ─── EventLedger — append-only, physically-ordered history with causal edges ──

// EventLedger never mutates or erases: a model rewind, an abandoned branch, and a superseded
// proposal are all retained. This is the minimal local store; enterprise ingestion consumes the
// same NDJSON. Mirrors events.py::EventLedger.
type EventLedger struct {
	events []RuntimeEvent
	byID   map[string]RuntimeEvent
}

// NewEventLedger builds an empty ledger.
func NewEventLedger() *EventLedger {
	return &EventLedger{byID: map[string]RuntimeEvent{}}
}

// Append records an event in physical order (never rewritten).
func (l *EventLedger) Append(e RuntimeEvent) RuntimeEvent {
	l.events = append(l.events, e)
	l.byID[e.EventID] = e
	return e
}

// Events returns the ledger in physical order.
func (l *EventLedger) Events() []RuntimeEvent { return append([]RuntimeEvent{}, l.events...) }

// Get returns the event with the given id (ok=false if absent).
func (l *EventLedger) Get(eventID string) (RuntimeEvent, bool) {
	e, ok := l.byID[eventID]
	return e, ok
}

// Children returns the events whose parent is eventID, in physical order.
func (l *EventLedger) Children(eventID string) []RuntimeEvent {
	var out []RuntimeEvent
	for _, e := range l.events {
		if e.ParentEventID == eventID {
			out = append(out, e)
		}
	}
	return out
}

// BySession returns the events in one session, in physical order.
func (l *EventLedger) BySession(sessionID string) []RuntimeEvent {
	var out []RuntimeEvent
	for _, e := range l.events {
		if e.SessionID == sessionID {
			out = append(out, e)
		}
	}
	return out
}

// ToNDJSON serializes the whole ledger, one event per line, in physical order.
func (l *EventLedger) ToNDJSON() string {
	lines := make([]string, len(l.events))
	for i, e := range l.events {
		lines[i] = e.ToNDJSON()
	}
	return strings.Join(lines, "\n")
}

// EventLedgerFromNDJSON replays a ledger from NDJSON text (blank lines skipped).
func EventLedgerFromNDJSON(text string) (*EventLedger, error) {
	led := NewEventLedger()
	for _, line := range strings.Split(text, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		ev, err := RuntimeEventFromNDJSON(line)
		if err != nil {
			return nil, err
		}
		led.Append(ev)
	}
	return led, nil
}

// SessionSummary is the session-summary/v1 record: count, distinct types (sorted), failures.
func (l *EventLedger) SessionSummary(sessionID string) map[string]any {
	evs := l.BySession(sessionID)
	typeSet := map[string]bool{}
	failed := 0
	for _, e := range evs {
		typeSet[string(e.EventType)] = true
		if e.ResultStatus == RsFailed {
			failed++
		}
	}
	types := make([]string, 0, len(typeSet))
	for t := range typeSet {
		types = append(types, t)
	}
	sort.Strings(types)
	return map[string]any{
		"schema_version": "session-summary/v1",
		"session_id":     sessionID,
		"event_count":    len(evs),
		"types":          stringsToAny(types),
		"failed":         failed,
	}
}

// ─── Python-compatible encoders (bit-identity with json.dumps / list repr) ───

// pyJSON reproduces Python's json.dumps(v, sort_keys=True) with its DEFAULT separators (", " and
// ": ") and ensure_ascii escaping — the exact bytes the event_id key and NDJSON depend on.
func pyJSON(v any) string {
	var sb strings.Builder
	writePyJSON(&sb, v)
	return sb.String()
}

func writePyJSON(sb *strings.Builder, v any) {
	switch x := v.(type) {
	case nil:
		sb.WriteString("null")
	case bool:
		if x {
			sb.WriteString("true")
		} else {
			sb.WriteString("false")
		}
	case string:
		writePyString(sb, x)
	case int:
		sb.WriteString(strconv.Itoa(x))
	case int64:
		sb.WriteString(strconv.FormatInt(x, 10))
	case float64:
		sb.WriteString(pyFloat(x))
	case json.Number:
		sb.WriteString(string(x)) // preserve the parsed token (int or fractional) verbatim
	case []any:
		sb.WriteByte('[')
		for i, e := range x {
			if i > 0 {
				sb.WriteString(", ")
			}
			writePyJSON(sb, e)
		}
		sb.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(x))
		for k := range x {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		sb.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				sb.WriteString(", ")
			}
			writePyString(sb, k)
			sb.WriteString(": ")
			writePyJSON(sb, x[k])
		}
		sb.WriteByte('}')
	default:
		// Fallback: route anything else through encoding/json (rare; payloads are plain JSON).
		b, _ := json.Marshal(x)
		sb.Write(b)
	}
}

// writePyString escapes a string as Python json.dumps (ensure_ascii=True): escape " \ and the C0
// controls with short escapes where Python does (\b \f \n \r \t), everything else non-ASCII as
// \uXXXX, printable ASCII verbatim.
func writePyString(sb *strings.Builder, s string) {
	sb.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			sb.WriteString(`\"`)
		case '\\':
			sb.WriteString(`\\`)
		case '\b':
			sb.WriteString(`\b`)
		case '\f':
			sb.WriteString(`\f`)
		case '\n':
			sb.WriteString(`\n`)
		case '\r':
			sb.WriteString(`\r`)
		case '\t':
			sb.WriteString(`\t`)
		default:
			if r < 0x20 || r > 0x7e {
				if r > 0xffff { // surrogate pair, as Python \uXXXX\uXXXX
					r -= 0x10000
					hi := 0xd800 + (r >> 10)
					lo := 0xdc00 + (r & 0x3ff)
					fmt.Fprintf(sb, `\u%04x\u%04x`, hi, lo)
				} else {
					fmt.Fprintf(sb, `\u%04x`, r)
				}
			} else {
				sb.WriteRune(r)
			}
		}
	}
	sb.WriteByte('"')
}

// pyFloat renders a float as Python's repr would for common values (shortest round-trip). Full
// parity with CPython's float repr across all inputs is out of scope; event payloads are plain data
// and rarely carry floats — integers, strings and nested maps dominate.
func pyFloat(f float64) string {
	return strconv.FormatFloat(f, 'g', -1, 64)
}

// pyListRepr reproduces Python's repr() of a list of strings: [] or ['a', 'b'] with single quotes,
// ", " separators, and backslash/quote escaping — the exact form embedded in the event_id key.
func pyListRepr(items []string) string {
	parts := make([]string, len(items))
	for i, s := range items {
		esc := strings.ReplaceAll(s, `\`, `\\`)
		esc = strings.ReplaceAll(esc, `'`, `\'`)
		parts[i] = "'" + esc + "'"
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

// ─── small conversion helpers ────────────────────────────────────────────────

func stringsToAny(ss []string) []any {
	out := make([]any, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}

func anyToStrings(v any) []string {
	arr, ok := v.([]any)
	if !ok {
		return []string{}
	}
	out := make([]string, 0, len(arr))
	for _, e := range arr {
		out = append(out, asString(e))
	}
	return out
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}

func asMap(v any) map[string]any {
	m, _ := v.(map[string]any)
	return m
}

func asInt(v any) int {
	switch n := v.(type) {
	case json.Number:
		i, _ := n.Int64()
		return int(i)
	case float64:
		return int(n)
	case int:
		return n
	}
	return 0
}

// isEmptyField mirrors Python's `not d.get(f)` truthiness for the Validate check: empty string,
// empty slice, empty map, nil, 0, and false all count as missing.
func isEmptyField(v any) bool {
	switch x := v.(type) {
	case nil:
		return true
	case string:
		return x == ""
	case []any:
		return len(x) == 0
	case map[string]any:
		return len(x) == 0
	case int:
		return x == 0
	case bool:
		return !x
	}
	return false
}
